File: python/stream_orderbook.py
#!/usr/bin/env python3
"""
Order Book Streaming Example — L2 and L4 Order Books

This example demonstrates how to stream order book data:
- L2 Book: Aggregated by price level (WebSocket or gRPC)
- L4 Book: Individual orders with order IDs (gRPC only)

L4 order book is CRITICAL for:
- Market making: Know exact queue position and order sizes
- Order flow analysis: Detect large orders and icebergs
- Optimal execution: See exactly what you're crossing
- Latency-sensitive trading: Lower latency than WebSocket

Setup:
    pip install hyperliquid-sdk

Usage:
    export ENDPOINT="https://your-endpoint.hype-mainnet.quiknode.pro/YOUR_TOKEN"
    python stream_orderbook.py
"""

import os
import sys
import time
from datetime import datetime
from typing import Dict, List, Any

from hyperliquid_sdk import GRPCStream, Stream, ConnectionState

ENDPOINT = os.environ.get("ENDPOINT")
if not ENDPOINT:
    print("Order Book Streaming Example")
    print("=" * 60)
    print()
    print("Usage:")
    print("  export ENDPOINT='https://YOUR-ENDPOINT.quiknode.pro/TOKEN'")
    print("  python stream_orderbook.py")
    sys.exit(1)


def timestamp():
    return datetime.now().strftime("%H:%M:%S.%f")[:-3]


# ═══════════════════════════════════════════════════════════════════════════════
# L2 ORDER BOOK (Aggregated Price Levels)
# ═══════════════════════════════════════════════════════════════════════════════

class L2BookTracker:
    """Track L2 order book state."""

    def __init__(self, coin: str):
        self.coin = coin
        self.bids: List[Dict[str, Any]] = []
        self.asks: List[Dict[str, Any]] = []
        self.last_update = 0.0

    def update(self, data: Dict[str, Any]) -> None:
        """Update book from L2 data."""
        self.bids = data.get("bids", [])
        self.asks = data.get("asks", [])
        self.last_update = time.time()

    def best_bid(self) -> tuple:
        """Return (price, size) of best bid."""
        if not self.bids:
            return (0.0, 0.0)
        bid = self.bids[0]
        if isinstance(bid, list):
            return (float(bid[0]), float(bid[1]))
        return (float(bid.get("px", 0)), float(bid.get("sz", 0)))

    def best_ask(self) -> tuple:
        """Return (price, size) of best ask."""
        if not self.asks:
            return (0.0, 0.0)
        ask = self.asks[0]
        if isinstance(ask, list):
            return (float(ask[0]), float(ask[1]))
        return (float(ask.get("px", 0)), float(ask.get("sz", 0)))

    def spread(self) -> float:
        """Return spread in price."""
        bid_px, _ = self.best_bid()
        ask_px, _ = self.best_ask()
        return ask_px - bid_px if bid_px and ask_px else 0.0

    def spread_bps(self) -> float:
        """Return spread in basis points."""
        bid_px, _ = self.best_bid()
        ask_px, _ = self.best_ask()
        if not bid_px or not ask_px:
            return 0.0
        mid = (bid_px + ask_px) / 2
        return (ask_px - bid_px) / mid * 10000

    def depth(self, levels: int = 5) -> Dict[str, float]:
        """Return total size at top N levels."""
        def sum_size(book: List) -> float:
            total = 0.0
            for level in book[:levels]:
                if isinstance(level, list) and len(level) >= 2:
                    total += float(level[1])
                elif isinstance(level, dict):
                    total += float(level.get("sz", 0))
            return total

        return {
            "bid_depth": sum_size(self.bids),
            "ask_depth": sum_size(self.asks),
        }

    def display(self) -> None:
        """Display current book state."""
        bid_px, bid_sz = self.best_bid()
        ask_px, ask_sz = self.best_ask()
        depth = self.depth()

        print(f"\n{self.coin} L2 Order Book")
        print("-" * 40)
        print(f"  Best Bid:   ${bid_px:>12,.2f} x {bid_sz:>10.4f}")
        print(f"  Best Ask:   ${ask_px:>12,.2f} x {ask_sz:>10.4f}")
        print(f"  Spread:     ${self.spread():>12,.2f} ({self.spread_bps():.2f} bps)")
        print(f"  Bid Depth:  {depth['bid_depth']:>23.4f} (top 5)")
        print(f"  Ask Depth:  {depth['ask_depth']:>23.4f} (top 5)")
        print(f"  Levels:     {len(self.bids)} bids, {len(self.asks)} asks")


def stream_l2_grpc():
    """Stream L2 order book via gRPC."""
    print("\n" + "=" * 60)
    print("L2 ORDER BOOK via gRPC")
    print("=" * 60)
    print()
    print("L2 book aggregates all orders at each price level.")
    print("n_sig_figs controls aggregation precision.")
    print()

    tracker = L2BookTracker("BTC")
    update_count = 0

    def on_l2(data: Dict[str, Any]):
        nonlocal update_count
        update_count += 1
        tracker.update(data)

        if update_count <= 3:
            tracker.display()

    stream = GRPCStream(ENDPOINT, reconnect=False)

    # n_sig_figs options:
    # - 5: Full precision (most levels)
    # - 4: Some aggregation
    # - 3: More aggregation (fewer levels, larger sizes)
    stream.l2_book("BTC", on_l2, n_sig_figs=5)

    print("Subscribing to BTC L2 book via gRPC...")
    print("-" * 60)

    stream.start()

    start = time.time()
    while update_count < 3 and time.time() - start < 15:
        time.sleep(0.1)

    stream.stop()
    print(f"\nReceived {update_count} L2 updates via gRPC")


def stream_l2_websocket():
    """Stream L2 order book via WebSocket."""
    print("\n" + "=" * 60)
    print("L2 ORDER BOOK via WebSocket")
    print("=" * 60)

    tracker = L2BookTracker("BTC")
    update_count = 0

    def on_l2(data: Dict[str, Any]):
        nonlocal update_count
        update_count += 1

        # WebSocket format: {"channel": "l2Book", "data": {...}}
        book = data.get("data", {})
        levels = book.get("levels", [[], []])

        # Convert to common format
        converted = {
            "bids": levels[0] if len(levels) > 0 else [],
            "asks": levels[1] if len(levels) > 1 else [],
        }
        tracker.update(converted)

        if update_count <= 3:
            tracker.display()

    stream = Stream(ENDPOINT, reconnect=False)
    stream.l2_book("BTC", on_l2)

    print("Subscribing to BTC L2 book via WebSocket...")
    print("-" * 60)

    stream.start()

    start = time.time()
    while update_count < 3 and time.time() - start < 15:
        time.sleep(0.1)

    stream.stop()
    print(f"\nReceived {update_count} L2 updates via WebSocket")


# ═══════════════════════════════════════════════════════════════════════════════
# L4 ORDER BOOK (Individual Orders) — CRITICAL FOR TRADING
# ═══════════════════════════════════════════════════════════════════════════════

class L4BookTracker:
    """Track L4 order book with individual orders."""

    def __init__(self, coin: str):
        self.coin = coin
        self.bids: Dict[str, Dict[str, Any]] = {}  # oid -> order
        self.asks: Dict[str, Dict[str, Any]] = {}  # oid -> order
        self.last_update = 0.0

    def update(self, data: Dict[str, Any]) -> None:
        """Update book from L4 data."""
        self.last_update = time.time()

        # Process bids
        for bid in data.get("bids", []):
            self._process_level(bid, self.bids, "bid")

        # Process asks
        for ask in data.get("asks", []):
            self._process_level(ask, self.asks, "ask")

    def _process_level(self, level: Any, book: Dict, side: str) -> None:
        """Process a single price level."""
        if isinstance(level, list):
            if len(level) >= 3:
                # Format: [price, size, order_id]
                px, sz, oid = level[0], level[1], str(level[2])
                if float(sz) == 0:
                    book.pop(oid, None)
                else:
                    book[oid] = {"px": px, "sz": sz, "oid": oid, "side": side}
            elif len(level) >= 2 and isinstance(level[1], list):
                # Format: [price, [[size, oid], ...]]
                px = level[0]
                for order in level[1]:
                    if isinstance(order, list) and len(order) >= 2:
                        sz, oid = order[0], str(order[1])
                        if float(sz) == 0:
                            book.pop(oid, None)
                        else:
                            book[oid] = {"px": px, "sz": sz, "oid": oid, "side": side}

    def get_sorted_bids(self) -> List[Dict]:
        """Return bids sorted by price (highest first)."""
        return sorted(
            self.bids.values(),
            key=lambda x: float(x["px"]),
            reverse=True
        )

    def get_sorted_asks(self) -> List[Dict]:
        """Return asks sorted by price (lowest first)."""
        return sorted(
            self.asks.values(),
            key=lambda x: float(x["px"])
        )

    def orders_at_price(self, price: float, side: str = "bid") -> List[Dict]:
        """Get all orders at a specific price."""
        book = self.bids if side == "bid" else self.asks
        return [
            order for order in book.values()
            if abs(float(order["px"]) - price) < 0.01  # Price tolerance
        ]

    def total_orders(self) -> tuple:
        """Return (bid_count, ask_count)."""
        return (len(self.bids), len(self.asks))

    def display(self, levels: int = 5) -> None:
        """Display current L4 book state."""
        sorted_bids = self.get_sorted_bids()
        sorted_asks = self.get_sorted_asks()

        print(f"\n{self.coin} L4 Order Book (Individual Orders)")
        print("=" * 60)

        # Group orders by price
        bid_prices: Dict[float, List[Dict]] = {}
        for bid in sorted_bids[:20]:  # Top 20 orders
            px = float(bid["px"])
            if px not in bid_prices:
                bid_prices[px] = []
            bid_prices[px].append(bid)

        ask_prices: Dict[float, List[Dict]] = {}
        for ask in sorted_asks[:20]:
            px = float(ask["px"])
            if px not in ask_prices:
                ask_prices[px] = []
            ask_prices[px].append(ask)

        # Display asks (top levels, reversed for display)
        print("\nASKS (top 5 price levels):")
        sorted_ask_prices = sorted(ask_prices.keys())[:levels]
        for px in reversed(sorted_ask_prices):
            orders = ask_prices[px]
            total_sz = sum(float(o["sz"]) for o in orders)
            print(f"  ${px:>12,.2f} | {total_sz:>10.4f} | {len(orders):>3} orders")
            # Show individual orders (first 3)
            for i, order in enumerate(orders[:3]):
                print(f"               └─ {float(order['sz']):>10.4f} (oid: {order['oid'][:8]}...)")

        print("  " + "-" * 56)
        print("  " + " " * 14 + "SPREAD")
        print("  " + "-" * 56)

        # Display bids (top levels)
        print("\nBIDS (top 5 price levels):")
        sorted_bid_prices = sorted(bid_prices.keys(), reverse=True)[:levels]
        for px in sorted_bid_prices:
            orders = bid_prices[px]
            total_sz = sum(float(o["sz"]) for o in orders)
            print(f"  ${px:>12,.2f} | {total_sz:>10.4f} | {len(orders):>3} orders")
            for i, order in enumerate(orders[:3]):
                print(f"               └─ {float(order['sz']):>10.4f} (oid: {order['oid'][:8]}...)")

        bid_count, ask_count = self.total_orders()
        print(f"\nTotal: {bid_count} bid orders, {ask_count} ask orders")


def stream_l4_book():
    """Stream L4 order book via gRPC (individual orders)."""
    print("\n" + "=" * 60)
    print("L4 ORDER BOOK via gRPC (Individual Orders)")
    print("=" * 60)
    print()
    print("L4 book shows EVERY individual order with order ID.")
    print("This is CRITICAL for market making and order flow analysis.")
    print()
    print("Use cases:")
    print("  - See exact queue position")
    print("  - Detect large orders / icebergs")
    print("  - Know exactly what you're crossing")
    print("  - Analyze order flow")
    print()

    tracker = L4BookTracker("BTC")
    update_count = 0

    def on_l4(data: Dict[str, Any]):
        nonlocal update_count
        update_count += 1
        tracker.update(data)

        if update_count <= 3:
            tracker.display(levels=3)
            print()

    stream = GRPCStream(ENDPOINT, reconnect=False)
    stream.l4_book("BTC", on_l4)

    print("Subscribing to BTC L4 book via gRPC...")
    print("-" * 60)

    stream.start()

    start = time.time()
    while update_count < 3 and time.time() - start < 15:
        time.sleep(0.1)

    stream.stop()
    print(f"\nReceived {update_count} L4 updates")


# ═══════════════════════════════════════════════════════════════════════════════
# COMPARISON: L2 vs L4 Order Book
# ═══════════════════════════════════════════════════════════════════════════════

def comparison():
    """Show comparison between L2 and L4 order books."""
    print("\n" + "=" * 60)
    print("L2 vs L4 ORDER BOOK COMPARISON")
    print("=" * 60)
    print()
    print("┌─────────────────────────────────────────────────────────────┐")
    print("│                    L2 ORDER BOOK                            │")
    print("├─────────────────────────────────────────────────────────────┤")
    print("│ • Aggregated by price level                                 │")
    print("│ • Shows total size at each price                            │")
    print("│ • Available via WebSocket AND gRPC                          │")
    print("│ • Lower bandwidth                                           │")
    print("│ • Good for: Price monitoring, simple trading                │")
    print("├─────────────────────────────────────────────────────────────┤")
    print("│ Example:                                                    │")
    print("│   Price: $95,000.00 | Total Size: 10.5 BTC                  │")
    print("│   (You don't know how many orders or their sizes)           │")
    print("└─────────────────────────────────────────────────────────────┘")
    print()
    print("┌─────────────────────────────────────────────────────────────┐")
    print("│                    L4 ORDER BOOK                            │")
    print("├─────────────────────────────────────────────────────────────┤")
    print("│ • Individual orders with order IDs                          │")
    print("│ • Shows each order separately                               │")
    print("│ • Available via gRPC ONLY (higher performance)              │")
    print("│ • Higher bandwidth but more detail                          │")
    print("│ • Good for: Market making, HFT, order flow analysis         │")
    print("├─────────────────────────────────────────────────────────────┤")
    print("│ Example:                                                    │")
    print("│   Price: $95,000.00                                         │")
    print("│     └─ Order 1: 5.0 BTC (oid: abc123...)                    │")
    print("│     └─ Order 2: 3.0 BTC (oid: def456...)                    │")
    print("│     └─ Order 3: 2.5 BTC (oid: ghi789...)                    │")
    print("│   (You see every order and can track queue position)        │")
    print("└─────────────────────────────────────────────────────────────┘")
    print()


# ═══════════════════════════════════════════════════════════════════════════════
# MAIN
# ═══════════════════════════════════════════════════════════════════════════════

if __name__ == "__main__":
    print("=" * 60)
    print("Order Book Streaming Examples")
    print("=" * 60)
    print(f"Endpoint: {ENDPOINT[:50]}...")

    try:
        # Show comparison
        comparison()

        # Run L2 examples
        stream_l2_grpc()
        stream_l2_websocket()

        # Run L4 example (CRITICAL)
        stream_l4_book()

        print()
        print("=" * 60)
        print("All examples completed!")
        print("=" * 60)

    except KeyboardInterrupt:
        print("\nDemo interrupted.")
    except Exception as e:
        print(f"\nError: {e}")
        raise

File: typescript/stream_orderbook.ts
#!/usr/bin/env npx ts-node
/**
 * Order Book Streaming Example — L2 and L4 Order Books via gRPC
 *
 * This example demonstrates how to stream order book data via gRPC:
 * - L2 Book: Aggregated by price level (total size and order count per price)
 * - L4 Book: Individual orders with order IDs
 *
 * Note: L2/L4 order books are only available via gRPC on QuickNode.
 *       WebSocket streaming provides book_updates (incremental deltas) instead.
 *
 * Use cases:
 * - L2 Book: Market depth, spread monitoring, analytics dashboards
 * - L4 Book: HFT, quant trading, market making, order flow analysis
 *
 * Setup:
 *     npm install hyperliquid-sdk @grpc/grpc-js @grpc/proto-loader
 *
 * Usage:
 *     export ENDPOINT="https://your-endpoint.hype-mainnet.quiknode.pro/YOUR_TOKEN"
 *     npx ts-node stream_orderbook.ts
 */

import { GRPCStream } from 'hyperliquid-sdk';

const ENDPOINT = process.env.ENDPOINT;
if (!ENDPOINT) {
  console.log("Order Book Streaming Example");
  console.log("=".repeat(60));
  console.log();
  console.log("Usage:");
  console.log("  export ENDPOINT='https://YOUR-ENDPOINT.quiknode.pro/TOKEN'");
  console.log("  npx ts-node stream_orderbook.ts");
  process.exit(1);
}

function timestamp(): string {
  const now = new Date();
  return now.toTimeString().split(' ')[0] + '.' + String(now.getMilliseconds()).padStart(3, '0');
}

// ═══════════════════════════════════════════════════════════════════════════════
// L2 ORDER BOOK (Aggregated Price Levels)
// ═══════════════════════════════════════════════════════════════════════════════

class L2BookTracker {
  coin: string;
  bids: unknown[] = [];
  asks: unknown[] = [];
  lastUpdate = 0;

  constructor(coin: string) {
    this.coin = coin;
  }

  update(data: Record<string, unknown>): void {
    this.bids = (data.bids || []) as unknown[];
    this.asks = (data.asks || []) as unknown[];
    this.lastUpdate = Date.now();
  }

  bestBid(): [number, number] {
    if (!this.bids.length) return [0, 0];
    const bid = this.bids[0];
    if (Array.isArray(bid)) return [parseFloat(String(bid[0])), parseFloat(String(bid[1]))];
    const b = bid as Record<string, unknown>;
    return [parseFloat(String(b.px || 0)), parseFloat(String(b.sz || 0))];
  }

  bestAsk(): [number, number] {
    if (!this.asks.length) return [0, 0];
    const ask = this.asks[0];
    if (Array.isArray(ask)) return [parseFloat(String(ask[0])), parseFloat(String(ask[1]))];
    const a = ask as Record<string, unknown>;
    return [parseFloat(String(a.px || 0)), parseFloat(String(a.sz || 0))];
  }

  spread(): number {
    const [bidPx] = this.bestBid();
    const [askPx] = this.bestAsk();
    return bidPx && askPx ? askPx - bidPx : 0;
  }

  spreadBps(): number {
    const [bidPx] = this.bestBid();
    const [askPx] = this.bestAsk();
    if (!bidPx || !askPx) return 0;
    const mid = (bidPx + askPx) / 2;
    return (askPx - bidPx) / mid * 10000;
  }

  depth(levels: number = 5): Record<string, number> {
    let bidDepth = 0;
    for (const level of this.bids.slice(0, levels)) {
      if (Array.isArray(level) && level.length >= 2) {
        bidDepth += parseFloat(String(level[1]));
      } else if (typeof level === 'object' && level !== null) {
        bidDepth += parseFloat(String((level as Record<string, unknown>).sz || 0));
      }
    }

    let askDepth = 0;
    for (const level of this.asks.slice(0, levels)) {
      if (Array.isArray(level) && level.length >= 2) {
        askDepth += parseFloat(String(level[1]));
      } else if (typeof level === 'object' && level !== null) {
        askDepth += parseFloat(String((level as Record<string, unknown>).sz || 0));
      }
    }

    return { bidDepth, askDepth };
  }

  display(): void {
    const [bidPx, bidSz] = this.bestBid();
    const [askPx, askSz] = this.bestAsk();
    const d = this.depth();

    console.log(`\n${this.coin} L2 Order Book`);
    console.log("-".repeat(40));
    console.log(`  Best Bid:   $${bidPx.toLocaleString().padStart(12)} x ${bidSz.toFixed(4).padStart(10)}`);
    console.log(`  Best Ask:   $${askPx.toLocaleString().padStart(12)} x ${askSz.toFixed(4).padStart(10)}`);
    console.log(`  Spread:     $${this.spread().toLocaleString().padStart(12)} (${this.spreadBps().toFixed(2)} bps)`);
    console.log(`  Bid Depth:  ${d.bidDepth.toFixed(4).padStart(23)} (top 5)`);
    console.log(`  Ask Depth:  ${d.askDepth.toFixed(4).padStart(23)} (top 5)`);
    console.log(`  Levels:     ${this.bids.length} bids, ${this.asks.length} asks`);
  }
}

async function streamL2Grpc(): Promise<void> {
  console.log("\n" + "=".repeat(60));
  console.log("L2 ORDER BOOK via gRPC");
  console.log("=".repeat(60));
  console.log();
  console.log("L2 book aggregates all orders at each price level.");
  console.log("nSigFigs controls aggregation precision.");
  console.log();

  const tracker = new L2BookTracker("BTC");
  let updateCount = 0;

  const onL2 = (data: Record<string, unknown>) => {
    updateCount++;
    tracker.update(data);

    if (updateCount <= 3) {
      tracker.display();
    }
  };

  const stream = new GRPCStream(ENDPOINT, { reconnect: false });

  // nSigFigs options:
  // - 5: Full precision (most levels)
  // - 4: Some aggregation
  // - 3: More aggregation (fewer levels, larger sizes)
  stream.l2Book("BTC", onL2, { nSigFigs: 5 });

  console.log("Subscribing to BTC L2 book via gRPC...");
  console.log("-".repeat(60));

  await stream.start();

  const start = Date.now();
  while (updateCount < 3 && Date.now() - start < 15000) {
    await new Promise(resolve => setTimeout(resolve, 100));
  }

  stream.stop();
  console.log(`\nReceived ${updateCount} L2 updates via gRPC`);
}

// ═══════════════════════════════════════════════════════════════════════════════
// L4 ORDER BOOK (Individual Orders) — CRITICAL FOR TRADING
// ═══════════════════════════════════════════════════════════════════════════════

interface OrderEntry {
  px: string;
  sz: string;
  oid: string;
  side: string;
}

class L4BookTracker {
  coin: string;
  bids: Map<string, OrderEntry> = new Map();
  asks: Map<string, OrderEntry> = new Map();
  lastUpdate = 0;

  constructor(coin: string) {
    this.coin = coin;
  }

  update(data: Record<string, unknown>): void {
    this.lastUpdate = Date.now();

    // Process bids
    for (const bid of (data.bids || []) as unknown[]) {
      this.processLevel(bid, this.bids, "bid");
    }

    // Process asks
    for (const ask of (data.asks || []) as unknown[]) {
      this.processLevel(ask, this.asks, "ask");
    }
  }

  private processLevel(level: unknown, book: Map<string, OrderEntry>, side: string): void {
    if (Array.isArray(level)) {
      if (level.length >= 3) {
        // Format: [price, size, order_id]
        const px = String(level[0]);
        const sz = String(level[1]);
        const oid = String(level[2]);
        if (parseFloat(sz) === 0) {
          book.delete(oid);
        } else {
          book.set(oid, { px, sz, oid, side });
        }
      } else if (level.length >= 2 && Array.isArray(level[1])) {
        // Format: [price, [[size, oid], ...]]
        const px = String(level[0]);
        for (const order of level[1]) {
          if (Array.isArray(order) && order.length >= 2) {
            const sz = String(order[0]);
            const oid = String(order[1]);
            if (parseFloat(sz) === 0) {
              book.delete(oid);
            } else {
              book.set(oid, { px, sz, oid, side });
            }
          }
        }
      }
    }
  }

  getSortedBids(): OrderEntry[] {
    return Array.from(this.bids.values()).sort(
      (a, b) => parseFloat(b.px) - parseFloat(a.px)
    );
  }

  getSortedAsks(): OrderEntry[] {
    return Array.from(this.asks.values()).sort(
      (a, b) => parseFloat(a.px) - parseFloat(b.px)
    );
  }

  totalOrders(): [number, number] {
    return [this.bids.size, this.asks.size];
  }

  display(levels: number = 5): void {
    const sortedBids = this.getSortedBids();
    const sortedAsks = this.getSortedAsks();

    console.log(`\n${this.coin} L4 Order Book (Individual Orders)`);
    console.log("=".repeat(60));

    // Group orders by price
    const bidPrices: Map<number, OrderEntry[]> = new Map();
    for (const bid of sortedBids.slice(0, 20)) {
      const px = parseFloat(bid.px);
      if (!bidPrices.has(px)) bidPrices.set(px, []);
      bidPrices.get(px)!.push(bid);
    }

    const askPrices: Map<number, OrderEntry[]> = new Map();
    for (const ask of sortedAsks.slice(0, 20)) {
      const px = parseFloat(ask.px);
      if (!askPrices.has(px)) askPrices.set(px, []);
      askPrices.get(px)!.push(ask);
    }

    // Display asks (top levels, reversed for display)
    console.log("\nASKS (top 5 price levels):");
    const sortedAskPrices = Array.from(askPrices.keys()).sort((a, b) => a - b).slice(0, levels);
    for (const px of sortedAskPrices.reverse()) {
      const orders = askPrices.get(px)!;
      const totalSz = orders.reduce((sum, o) => sum + parseFloat(o.sz), 0);
      console.log(`  $${px.toLocaleString().padStart(12)} | ${totalSz.toFixed(4).padStart(10)} | ${orders.length.toString().padStart(3)} orders`);
      // Show individual orders (first 3)
      for (const order of orders.slice(0, 3)) {
        console.log(`               └─ ${parseFloat(order.sz).toFixed(4).padStart(10)} (oid: ${order.oid.slice(0, 8)}...)`);
      }
    }

    console.log("  " + "-".repeat(56));
    console.log("  " + " ".repeat(14) + "SPREAD");
    console.log("  " + "-".repeat(56));

    // Display bids (top levels)
    console.log("\nBIDS (top 5 price levels):");
    const sortedBidPrices = Array.from(bidPrices.keys()).sort((a, b) => b - a).slice(0, levels);
    for (const px of sortedBidPrices) {
      const orders = bidPrices.get(px)!;
      const totalSz = orders.reduce((sum, o) => sum + parseFloat(o.sz), 0);
      console.log(`  $${px.toLocaleString().padStart(12)} | ${totalSz.toFixed(4).padStart(10)} | ${orders.length.toString().padStart(3)} orders`);
      for (const order of orders.slice(0, 3)) {
        console.log(`               └─ ${parseFloat(order.sz).toFixed(4).padStart(10)} (oid: ${order.oid.slice(0, 8)}...)`);
      }
    }

    const [bidCount, askCount] = this.totalOrders();
    console.log(`\nTotal: ${bidCount} bid orders, ${askCount} ask orders`);
  }
}

async function streamL4Book(): Promise<void> {
  console.log("\n" + "=".repeat(60));
  console.log("L4 ORDER BOOK via gRPC (Individual Orders)");
  console.log("=".repeat(60));
  console.log();
  console.log("L4 book shows EVERY individual order with order ID.");
  console.log("This is CRITICAL for market making and order flow analysis.");
  console.log();
  console.log("Use cases:");
  console.log("  - See exact queue position");
  console.log("  - Detect large orders / icebergs");
  console.log("  - Know exactly what you're crossing");
  console.log("  - Analyze order flow");
  console.log();

  const tracker = new L4BookTracker("BTC");
  let updateCount = 0;

  const onL4 = (data: Record<string, unknown>) => {
    updateCount++;
    tracker.update(data);

    if (updateCount <= 3) {
      tracker.display(3);
      console.log();
    }
  };

  const stream = new GRPCStream(ENDPOINT, { reconnect: false });
  stream.l4Book("BTC", onL4);

  console.log("Subscribing to BTC L4 book via gRPC...");
  console.log("-".repeat(60));

  await stream.start();

  const start = Date.now();
  while (updateCount < 3 && Date.now() - start < 15000) {
    await new Promise(resolve => setTimeout(resolve, 100));
  }

  stream.stop();
  console.log(`\nReceived ${updateCount} L4 updates`);
}

// ═══════════════════════════════════════════════════════════════════════════════
// COMPARISON: L2 vs L4 Order Book
// ═══════════════════════════════════════════════════════════════════════════════

function comparison(): void {
  console.log("\n" + "=".repeat(60));
  console.log("L2 vs L4 ORDER BOOK COMPARISON");
  console.log("=".repeat(60));
  console.log();
  console.log("┌─────────────────────────────────────────────────────────────┐");
  console.log("│                    L2 ORDER BOOK                            │");
  console.log("├─────────────────────────────────────────────────────────────┤");
  console.log("│ • Aggregated by price level                                 │");
  console.log("│ • Shows total size at each price                            │");
  console.log("│ • Available via gRPC (StreamL2Book)                         │");
  console.log("│ • Lower bandwidth                                           │");
  console.log("│ • Good for: Price monitoring, simple trading                │");
  console.log("├─────────────────────────────────────────────────────────────┤");
  console.log("│ Example:                                                    │");
  console.log("│   Price: $95,000.00 | Total Size: 10.5 BTC                  │");
  console.log("│   (You don't know how many orders or their sizes)           │");
  console.log("└─────────────────────────────────────────────────────────────┘");
  console.log();
  console.log("┌─────────────────────────────────────────────────────────────┐");
  console.log("│                    L4 ORDER BOOK                            │");
  console.log("├─────────────────────────────────────────────────────────────┤");
  console.log("│ • Individual orders with order IDs                          │");
  console.log("│ • Shows each order separately                               │");
  console.log("│ • Available via gRPC (StreamL4Book)                         │");
  console.log("│ • Higher bandwidth but more detail                          │");
  console.log("│ • Good for: Market making, HFT, order flow analysis         │");
  console.log("├─────────────────────────────────────────────────────────────┤");
  console.log("│ Example:                                                    │");
  console.log("│   Price: $95,000.00                                         │");
  console.log("│     └─ Order 1: 5.0 BTC (oid: abc123...)                    │");
  console.log("│     └─ Order 2: 3.0 BTC (oid: def456...)                    │");
  console.log("│     └─ Order 3: 2.5 BTC (oid: ghi789...)                    │");
  console.log("│   (You see every order and can track queue position)        │");
  console.log("└─────────────────────────────────────────────────────────────┘");
  console.log();
  console.log("Note: For incremental book changes (deltas), use WebSocket:");
  console.log("  stream.bookUpdates(coins, callback)");
  console.log();
}

// ═══════════════════════════════════════════════════════════════════════════════
// MAIN
// ═══════════════════════════════════════════════════════════════════════════════

async function main() {
  console.log("=".repeat(60));
  console.log("Order Book Streaming Examples (gRPC)");
  console.log("=".repeat(60));
  console.log(`Endpoint: ${ENDPOINT.slice(0, 50)}...`);

  try {
    // Show comparison
    comparison();

    // Run L2 example
    await streamL2Grpc();

    // Run L4 example (CRITICAL)
    await streamL4Book();

    console.log();
    console.log("=".repeat(60));
    console.log("All examples completed!");
    console.log("=".repeat(60));

  } catch (e) {
    if (e instanceof Error && e.message.includes('interrupt')) {
      console.log("\nDemo interrupted.");
    } else {
      console.log(`\nError: ${e}`);
      throw e;
    }
  }
}

main().catch(console.error);
